[LoopVectorize] Further improve cost model for early exit loops

david-arm opened the pull request:
Following on from #125058, this patch takes into account the
work done in the vector early exit block when assessing the
profitability of vectorising the loop. I have renamed
areRuntimeChecksProfitable to isOutsideLoopWorkProfitable and
we now pass in the early exit costs. As part of this, I have
added the ExtractFirstActive opcode to VPInstruction::computeCost.

It's worth pointing out that when we assess profitability of the
loop we calculate a minimum trip count and compare that against
the *maximum* trip count. However, since the loop has an early
exit the runtime trip count can still end up being less than the
minimum. Alternatively, we may never take the early exit at all
at runtime and so we have the opposite problem of over-estimating
the cost of the loop. The loop vectoriser cannot simultaneously
take two contradictory positions and so I feel the only sensible
thing to do is be conservative and assume the loop will be more
expensive than loops without early exits.

We may find in future that we need to adjust the cost according to
the probability of taking the early exit. This will become even
more important once we support multiple early exits. However, we
have to start somewhere and we can always revisit this later.

david-arm committed:
[LoopVectorize] Further improve cost model for early exit loops

Following on from #125058, this patch takes into account the
work done in the vector early exit block when assessing the
profitability of vectorising the loop. I have renamed
areRuntimeChecksProfitable to isOutsideLoopWorkProfitable and
we now pass in the early exit costs. As part of this, I have
added the ExtractFirstActive opcode to VPInstruction::computeCost.

It's worth pointing out that when we assess profitability of the
loop we calculate a minimum trip count and compare that against
the *maximum* trip count. However, since the loop has an early
exit the runtime trip count can still end up being less than the
minimum. Alternatively, we may never take the early exit at all
at runtime and so we have the opposite problem of over-estimating
the cost of the loop. The loop vectoriser cannot simultaneously
take two contradictory positions and so I feel the only sensible
thing to do is be conservative and assume the loop will be more
expensive than loops without early exits.

We may find in future that we need to adjust the cost according to
the probability of taking the early exit. This will become even
more important once we support multiple early exits. However, we
have to start somewhere and we can always revisit this later.

david-arm commented:
Gentle ping. :)

MacDue commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10077,19 +10077,46 @@ static void checkMixedPrecision(Loop *L, OptimizationRemarkEmitter *ORE) {
+  LLVM_DEBUG(
+      dbgs() << "Calculating cost of work in vector early exit block:\n");
+  for (auto *ExitVPBB : Plan.getExitBlocks()) {
+    for (auto *PredVPBB : ExitVPBB->getPredecessors())
+      if (PredVPBB != Plan.getMiddleBlock())
```
Could you add a comment explaining this step? I was not clear until I found: https://llvm.org/docs/Vectorizers.html#early-exit-vectorization, where it seems like this is looking for `vector.early.exit` from `early.exit` (maybe just link to this diagram).

david-arm replied:
Done, hope the comments explain it better!

MacDue replied:
Looks good, thanks :+1:

david-arm committed:
Address review comments

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10077,19 +10077,46 @@ static void checkMixedPrecision(Loop *L, OptimizationRemarkEmitter *ORE) {
-                                       PredicatedScalarEvolution &PSE,
-                                       ScalarEpilogueLowering SEL,
-                                       std::optional<unsigned> VScale) {
+static InstructionCost calculateEarlyExitCost(LoopVectorizationCostModel &CM,
```
Done

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10077,19 +10077,46 @@ static void checkMixedPrecision(Loop *L, OptimizationRemarkEmitter *ORE) {
+  return Cost;
+}
+
+static bool isOutsideLoopWorkProfitable(GeneratedRTChecks &Checks,
```
Done

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10132,7 +10159,7 @@ static bool areRuntimeChecksProfitable(GeneratedRTChecks &Checks,
   // is rounded up, hence we get an upper estimate of the TC.
   unsigned IntVF = getEstimatedRuntimeVF(VF.Width, VScale);
-  uint64_t RtC = *CheckCost.getValue();
+  uint64_t RtC = *TotalCost.getValue();
```
Done

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -726,6 +726,19 @@ InstructionCost VPInstruction::computeCost(ElementCount VF,
+    auto *PredTy = toVectorTy(Ctx.Types.inferScalarType(getOperand(1)), VF);
+    IntrinsicCostAttributes Attrs(
+        Intrinsic::experimental_cttz_elts, Type::getInt64Ty(Ctx.LLVMCtx),
+        {PoisonValue::get(PredTy), ConstantInt::getTrue(Ctx.LLVMCtx)});
```
Yep, you're right. Good suggestion!

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/early_exit_costs.ll:
```diff
@@ -0,0 +1,47 @@
+
+declare void @init_mem(ptr, i64);
+
+define i64 @same_exit_block_pre_inc_use1() #1 {
```
Done

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10564,12 +10591,17 @@ bool LoopVectorizePass::processLoop(Loop *L) {
       Checks.create(L, *LVL.getLAI(), PSE.getPredicate(), VF.Width, SelectedIC);
 
+    InstructionCost EarlyExitCost = InstructionCost::getInvalid();
+    if (VF.Width.isVector() && LVL.hasUncountableEarlyExit())
```
Done

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10564,12 +10591,17 @@ bool LoopVectorizePass::processLoop(Loop *L) {
+    InstructionCost EarlyExitCost = InstructionCost::getInvalid();
+    if (VF.Width.isVector() && LVL.hasUncountableEarlyExit())
+      EarlyExitCost =
+          calculateEarlyExitCost(CM, LVP.getPlanFor(VF.Width), VF.Width);
```
Done

david-arm commented:
Gentle ping. :)

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10165,19 +10165,59 @@ static void checkMixedPrecision(Loop *L, OptimizationRemarkEmitter *ORE) {
+  // Add on the cost of work required in the vector early exit block, if one
+  // exists.
+  if (CM.Legal->hasUncountableEarlyExit())
+    TotalCost += calculateEarlyExitCost(CM, Plan, VF.Width);
```
Can this be done unconditionally so we don't have to pull in a dependency on LoopVectorizationCostModel? Would be good if we could avoid adding new uses, to not make it harder to remove.

david-arm replied:
Done. Note I had to update some tests because of the extra debug output.

david-arm committed:
Address review comments

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10165,19 +10165,59 @@ static void checkMixedPrecision(Loop *L, OptimizationRemarkEmitter *ORE) {
+      // values of variables used outside the loop.
+      if (PredVPBB != Plan.getMiddleBlock())
+        for (auto &R : *(cast<VPBasicBlock>(PredVPBB)))
+          Cost += R.cost(VF, CostCtx);
```
Done

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -739,6 +739,19 @@ InstructionCost VPInstruction::computeCost(ElementCount VF,
+    auto *VecTy = toVectorTy(Ctx.Types.inferScalarType(getOperand(0)), VF);
+    Cost += Ctx.TTI.getVectorInstrCost(Instruction::ExtractElement, VecTy,
+                                       Ctx.CostKind);
+    return Cost;
```
Done

david-arm committed:
Fix test after rebase

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10258,7 +10298,7 @@ static bool areRuntimeChecksProfitable(GeneratedRTChecks &Checks,
   // Skip vectorization if the expected trip count is less than the minimum
   // required trip count.
-  if (auto ExpectedTC = getSmallBestKnownTC(PSE, L)) {
+  if (auto ExpectedTC = getSmallBestKnownTC(PSE, CM.TheLoop)) {
```
unrelated?

david-arm replied:
Again, it is related because we don't have a `L` function argument.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10230,8 +10270,8 @@ static bool areRuntimeChecksProfitable(GeneratedRTChecks &Checks,
   // is rounded up, hence we get an upper estimate of the TC.
-  unsigned IntVF = getEstimatedRuntimeVF(VF.Width, VScale);
-  uint64_t RtC = *CheckCost.getValue();
+  unsigned IntVF = getEstimatedRuntimeVF(VF.Width, CM.getVScaleForTuning());
```
Unrelated?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10659,8 +10699,8 @@ bool LoopVectorizePass::processLoop(Loop *L) {
     if (!ForceVectorization &&
-        !areRuntimeChecksProfitable(Checks, VF, L, PSE, SEL,
-                                    CM.getVScaleForTuning())) {
+        !isOutsideLoopWorkProfitable(Checks, VF, CM, PSE,
```
passing CM is now unnecessary/unrelated?

david-arm replied:
Oh I read this after reading your 'unrelated' comments above, which I found confusing at first. The problem is that so many of the objects required to calculate the early exit cost live in the cost model class so it seemed simpler to pass that in here. Perhaps I'm missing something, but it wasn't clear from your comments above what you wanted me to do here? I thought originally you were suggesting I move the creation of the `VPCostContext` object into `isOutsideLoopWorkProfitable`, but are you suggesting I create the `VPCostContext` class here instead and pass it to `isOutsideLoopWorkProfitable`?

fhahn replied:
Adding here was my thinking, to avoid spreading further references of the legacy CM.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10175,19 +10175,57 @@ static void checkMixedPrecision(Loop *L, OptimizationRemarkEmitter *ORE) {
+  InstructionCost Cost = 0;
+  VPCostContext CostCtx(CM.TTI, *CM.TLI, CM.Legal->getWidestInductionType(), CM,
+                        CM.CostKind);
+  LLVM_DEBUG(
+      dbgs() << "Calculating cost of work in vector early exit block:\n");
```
Can we print this only if there's an early exit block?

david-arm replied:
We can, but then we want to print it before calling the cost function otherwise it would look weird with debug output like:

Cost of ...
Calculating cost of work in vector early exit block

So the only way to do this is to duplicate the loop below and check for an early exit. I can collect all the blocks in a vector and check for non-zero size I suppose. This debug printing issue is actually the reason I used to guard calling this function by querying Legal->hasUncountableEarlyExit(), but now that's been removed we unconditionally call this function.

fhahn replied:
Could the  print be moved to

```
      if (PredVPBB != Plan.getMiddleBlock())
         Cost += PredVPBB->cost(VF, CostCtx);
```
?

Currently the wording implies there's a single block, but could also include the name of the earlier exit block, as it should work for any number of exit blocks?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10175,19 +10175,57 @@ static void checkMixedPrecision(Loop *L, OptimizationRemarkEmitter *ORE) {
+static InstructionCost calculateEarlyExitCost(LoopVectorizationCostModel &CM,
+                                              VPlan &Plan, ElementCount VF) {
+  InstructionCost Cost = 0;
+  VPCostContext CostCtx(CM.TTI, *CM.TLI, CM.Legal->getWidestInductionType(), CM,
```
Could we pass the cost context instead of explicitly threading through the legacy cost model, to limit new uses?

david-arm commented:
Hi @fhahn, could you clarify what you mean by passing in `CostCtx`? I have addressed most of your review comments downstream, but it's not clear whether you want me to just pass it into `calculateEarlyExitCost` and create it in `isOutsideLoopWorkProfitable`, or pass it to `isOutsideLoopWorkProfitable`. Once I know your preference I can upload a new version - thanks!

fhahn reviewed (commented):
> Hi @fhahn, could you clarify what you mean by passing in `CostCtx`? I have addressed most of your review comments downstream, but it's not clear whether you want me to just pass it into `calculateEarlyExitCost` and create it in `isOutsideLoopWorkProfitable`, or pass it to `isOutsideLoopWorkProfitable`. Once I know your preference I can upload a new version - thanks!

My thinking was pass to `isOutsideLoopWorkProfitable` instead of the legacy cost model, to avoid spreading new references of it

david-arm committed:
Address review comments

david-arm commented:
> My thinking was pass to isOutsideLoopWorkProfitable instead of the legacy cost model, to avoid spreading new references of it

Perfect, thanks for clarifying!

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/low_trip_memcheck_cost.ll:
```diff
@@ -8,7 +8,7 @@ define void @no_outer_loop(ptr nocapture noundef %a, ptr nocapture noundef reado
 ; CHECK-NOT:  We expect runtime memory checks to be hoisted out of the outer loop.
 ; CHECK:      Total cost of runtime checks: 4
-; CHECK-NEXT: LV: Minimum required TC for runtime checks to be profitable:16
+; CHECK:      LV: Minimum required TC for runtime checks to be profitable:16
```
Those changes shouldn't be needed with the latest changes I think?

fhahn reviewed (approved):
LGTM, thanks.


Please strip the changes in `llvm/test/Transforms/LoopVectorize/AArch64/low_trip_memcheck_cost.ll` if they aren't needed, which I would expect with the latest changes

david-arm committed:
Address review comment

david-arm commented:
Linux builds seem to be stuck, but the Windows build has passed (and I ran make check-all downstream) so I'll merge it.
